pkg/bindings/stan/message.go

```diff
@@ -0,0 +1,98 @@
+package stan
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"github.com/cloudevents/sdk-go/pkg/binding"
+	"github.com/cloudevents/sdk-go/pkg/binding/format"
+	"github.com/cloudevents/sdk-go/pkg/binding/spec"
+	"github.com/nats-io/stan.go"
+)
+
+type Message struct {
+	*stan.Msg
+	ManualAcks bool
+}
+
+type MessageOption func(*Message) error
+
+func WithManualAcks() MessageOption {
+	return func(m *Message) error {
+		m.ManualAcks = true
+		return nil
+	}
+}
+
+// By the specification, all NATS message payloads must be in the JSON event format.
+// If this isn't the case then this function will return an error
+func NewMessage(msg *stan.Msg, opts ...MessageOption) (*Message, error) {
+	raw := make(map[string]json.RawMessage)
+
+	if err := json.Unmarshal(msg.Data, &raw); err != nil {
+		return nil, err
+	}
+
+	if _, err := spec.VS.FindVersion(versionExtractor(raw)); err != nil {
+		return nil, err
+	}
+
+	m := &Message{Msg: msg}
+
+	if err := m.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+func (m *Message) applyOptions(opts ...MessageOption) error {
+	for _, fn := range opts {
+		err := fn(m)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// All NATS messages are structured
+func (m *Message) Encoding() binding.Encoding {
+	return binding.EncodingStructured
+}
+
+func (m *Message) Structured(ctx context.Context, encoder binding.StructuredEncoder) error {
+	return encoder.SetStructuredEvent(ctx, format.JSON, bytes.NewReader(m.Data))
+}
+
+func (m *Message) Binary(context.Context, binding.BinaryEncoder) error {
+	return BinaryEncodingNotSupported
```

Comment:
By contract you should return `binding.ErrNotBinary`, not a custom error message